### PR TITLE
win: fix unexpected ECONNRESET error on TCP socket

### PR DESCRIPTION
**Pull request by twose**
Introduced by #3036.

The order of calling `uv_read_stop()` and `uv__tcp_try_cancel_reqs()` doesn't seem to matter. But calling `uv_read_stop()` first is more in line with the order in which the user (me) uses these APIs, and `uv_read_stop()` is more soft I think.

Close the tcp handle after `uv_read_start()` immediately without a timer can also reproduce the problem, but I still constructed a test that is closer to my example to make sure it doesn't go wrong again.

@vtjnash Thank you very much for your pointers, after spending some time reading the related source code, I also learned something, especially since I almost never read the source code under the Windows platform before.

Please correct me if I am doing something wrong :)

### CODE REVIEW

File: test/test-tcp-close-after-read-timeout.c [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+static void on_client_read(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf);
+static void on_client_timeout(uv_timer_t* handle);
+
+static void on_close(uv_handle_t* handle);
+
+
+static void on_client_connect(uv_connect_t* conn_req, int status) {
+  int r;
+
+  r = uv_read_start((uv_stream_t*)&client, on_client_alloc, on_client_read);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:46]:
```suggestion
  r = uv_read_start((uv_stream_t*) &client, on_client_alloc, on_client_read);
```

File: test/test-tcp-close-after-read-timeout.c [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  read_cb_called++;
+}
+
+
+static void on_client_timeout(uv_timer_t* handle) {
+  ASSERT_EQ(handle, &timer);
+  ASSERT_EQ(read_cb_called, 0);
+  uv_read_stop((uv_stream_t*)&client);
+  uv_close((uv_handle_t*)&client, on_close);
+  uv_close((uv_handle_t*)&timer, on_close);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:72]:
```suggestion
  uv_read_stop((uv_stream_t*) &client);
  uv_close((uv_handle_t*) &client, on_close);
  uv_close((uv_handle_t*) &timer, on_close);
```

File: test/test-tcp-close-after-read-timeout.c [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  static char slab[8];
+  buf->base = slab;
+  buf->len = sizeof(slab);
+}
+
+
+static void on_connection_read(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT_EQ(nread, UV_EOF);
+  read_cb_called++;
+  uv_close((uv_handle_t*)stream, on_close);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:88]:
```suggestion
  uv_close((uv_handle_t*) stream, on_close);
```

File: test/test-tcp-close-after-read-timeout.c [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+
+static void on_connection(uv_stream_t* server, int status) {
+  int r;
+
+  ASSERT_EQ(status, 0);
+  ASSERT_EQ(uv_accept(server, (uv_stream_t*)&connection), 0);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:96]:
```suggestion
  ASSERT_EQ(uv_accept(server, (uv_stream_t*) &connection), 0);
```

File: test/test-tcp-close-after-read-timeout.c [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+  r = uv_read_start((uv_stream_t*)&connection, on_connection_alloc, on_connection_read);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:98]:
```suggestion
  r = uv_read_start((uv_stream_t*) &connection, on_connection_alloc, on_connection_read);
```

File: test/test-tcp-close-after-read-timeout.c [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  ASSERT_EQ(r, 0);
+}
+
+
+static void on_close(uv_handle_t* handle) {
+  ASSERT(handle == (uv_handle_t*)&client || handle == (uv_handle_t*)&connection || handle == (uv_handle_t*)&timer);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:104]: style nit: wrap this at 80 characters
```suggestion
  ASSERT(handle == (uv_handle_t*) &client || handle == (uv_handle_t*) &connection || handle == (uv_handle_t*) &timer);
```

File: test/test-tcp-close-after-read-timeout.c (hunk 5 above)
Comment (vtjnash) [reply test/test-tcp-close-after-read-timeout.c:98]: style nit: wrap this at 80 characters

File: test/test-tcp-close-after-read-timeout.c [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+  ASSERT_EQ(uv_ip4_addr("127.0.0.1", TEST_PORT, &addr), 0);
+
+  r = uv_tcp_init(loop, handle);
+  ASSERT_EQ(r, 0);
+
+  r = uv_tcp_bind(handle, (const struct sockaddr*) &addr, 0);
+  ASSERT_EQ(r, 0);
+
+  r = uv_listen((uv_stream_t*)handle, 128, on_connection);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:121]:
```suggestion
  r = uv_listen((uv_stream_t*) handle, 128, on_connection);
```

File: test/test-tcp-close-after-read-timeout.c [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  ASSERT_EQ(r, 0);
+
+  uv_unref((uv_handle_t*)handle);
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:124]:
```suggestion
  uv_unref((uv_handle_t*) handle);
```

**Review by vtjnash — APPROVED** [review]
Seems good to me. Surprisingly straightforward to fix.

Can you line wrap the test to 80 characters? That's the main thing to fix, and I only marked a couple lines

File: test/test-tcp-close-after-read-timeout.c [hunk 9]
```diff
@@ -0,0 +1,170 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (vtjnash) [inline test/test-tcp-close-after-read-timeout.c:1]:
```suggestion
/* Copyright libuv project and contributors. All rights reserved.
```

**Comment by twose** [comment]
> Seems good to me. Surprisingly straightforward to fix.

Also surprised...

> Can you line wrap the test to 80 characters? That's the main thing to fix, and I only marked a couple lines

Fixed. It seems that the code style in the existing tests is not unified, I just found one to imitate. IMO, 80 characters seems a bit conservative nowadays. XD

**Comment by vtjnash** [comment]
What can I say haha? libuv is a conservative library.

**Pull request merged**